### PR TITLE
Password history policy - add time based policy

### LINKED ISSUE #30211: Password history policy - add time based policy.
### Description

Right now keycloak supports one password history policy - depending on configuration, when the password is changed it can't be reused. We have an option to configure a number of passwords stored, if the password is too recent - the policy will kick in. 
Feature enhancement would be to add an option to define a period instead of the number of uses. Let's say we want to make sure that users can't reuse the same password in a certain number of days.
The time-based policy would work with the current password history policy, more restrictive policy would be preferred if both are set.

**Naming used in examples:**
- history policy = current implementation, a user can't reuse password X times
- time based policy = approach that calculates password age. To reuse, the password has to be older than X days.

**Scenario 1:**
Realm configuration 
- time-based policy set to 30 days
- history policy set to 5 passwords

The user recently changed the passwords:
- password_1 35 days ago
- password_2 25 days ago
- password_3 15 days ago
- password_4 10 days ago

When the user tries to reuse password_1 the history policy kicks in, a new password should be different.

**Scenario 2:**
Realm configuration 

- time-based policy set to 30 days
- history policy set to 2 passwords

The user recently changed the passwords:
- password_1 25 days ago
- password_2 20 days ago
- password_3 15 days ago
- password_4 10 days ago

When the user tries to reuse password_1 the time-based policy kicks in, a new password should be different.

**Scenario 3:**
Realm configuration 
- time-based policy set to 30 days
- history policy set to 2 passwords

The user recently changed the passwords:
- password_1 35 days ago
- password_2 20 days ago
- password_3 15 days ago
- password_4 10 days ago

When the user tries to reuse password_1 it's possible as both policies allow it. If the user tries password_2, a time-based policy will kick in not allowing the password to be reused. If the user tries password_3 and password_4 both policies wouldn't validate the change.

### Discussion

https://github.com/keycloak/keycloak/discussions/30210

### Motivation

The proposed password policy often exists in various authentication systems. Password can't be reused for a certain amount of time.

### Details

UI example of implemented change (both password history policies are set)
image
UI example of working policy (user tries to reuse password)
image

### PR DESCRIPTION
**Pull request by MaciejMierzwa**
Issue: https://github.com/keycloak/keycloak/issues/30211
Discussion: https://github.com/keycloak/keycloak/discussions/30210

Closes https://github.com/keycloak/keycloak/issues/30211

### CODE REVIEW

**Comment by MaciejMierzwa** [comment]
Implementation is based on other password history based policy - [HistoryPasswordPolicyProvider.java](https://github.com/keycloak/keycloak/blob/main/server-spi-private/src/main/java/org/keycloak/policy/HistoryPasswordPolicyProvider.java)

File: common/src/main/java/org/keycloak/common/util/Time.java
Comment (thomasdarimont) [inline common/src/main/java/org/keycloak/common/util/Time.java:77]:
```suggestion
```

File: server-spi-private/src/main/java/org/keycloak/policy/AgePasswordPolicyProvider.java
Comment (thomasdarimont) [inline server-spi-private/src/main/java/org/keycloak/policy/AgePasswordPolicyProvider.java:16]:
```suggestion
    private static final String ERROR_MESSAGE = "invalidPasswordGenericMessage";
```

File: server-spi-private/src/main/java/org/keycloak/policy/AgePasswordPolicyProvider.java
Comment (thomasdarimont) [inline server-spi-private/src/main/java/org/keycloak/policy/AgePasswordPolicyProvider.java:18]:
```suggestion
    private final KeycloakSession session;
```

File: server-spi/src/main/java/org/keycloak/models/PasswordPolicy.java
Comment (thomasdarimont) [inline server-spi/src/main/java/org/keycloak/models/PasswordPolicy.java:103]:
```suggestion
        if (policyConfig.containsKey(PASSWORD_AGE)) {
```

**Review by thomasdarimont — CHANGES_REQUESTED** [review]
Just a few nitpicks - otherwise LGTM

File: server-spi/src/main/java/org/keycloak/models/PasswordPolicy.java
Comment (MaciejMierzwa) [reply server-spi/src/main/java/org/keycloak/models/PasswordPolicy.java]: Updated branch with suggested changes. I also increased the time offset in tests, it was set to a couple of seconds, but now it is a couple of minutes. It should help with test flakiness in case of longer runs.

**Comment by MaciejMierzwa** [comment]
Updated branch with suggested changes. I also increased the time offset in tests, it was set to a couple of seconds, but now it is a couple of minutes. It should help with test flakiness in case of longer runs.

File: common/src/main/java/org/keycloak/common/util/Time.java
Comment (mabartos) [inline common/src/main/java/org/keycloak/common/util/Time.java:78]: I've just noticed that a small improvement might be made for this. I'll leave the rest of the review up to @keycloak/core team.

```suggestion
        return Duration.ofDays(days).toMillis();
```

Comment (MaciejMierzwa) [reply common/src/main/java/org/keycloak/common/util/Time.java:78]: Nice, I think I'll just remove the method altogether.

**Review by sguilhen — COMMENTED** [review]
LGTM, although it is important to get a review from core-clients. Only thing I'm missing is the license header in the new files.

**Review by pedroigor — COMMENTED** [review]
Shall we add this one to https://www.keycloak.org/docs/latest/server_admin/#_password-policies?

**Comment by MaciejMierzwa** [comment]
Ok, I checked the license header and authors in other files, and added it to the new files. If I used the wrong template, or we shouldn't add authors anymore, let me know.

**Comment by MaciejMierzwa** [comment]
Ok, I see the license header and authors in other files. I've found it's not uniform across the project, let me know if I've used a wrong template.
Also added a short description of new policy in docs, in place highlighted by [pedroigor](https://github.com/pedroigor): https://github.com/keycloak/keycloak/pull/30215#pullrequestreview-2113221242

**Comment by sguilhen** [comment]
@pedroigor can you trigger the CI?

**Comment by codespearhead** [comment]
@MaciejMierzwa Can you [link this PR to that issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) so that the issue is automatically closed when this PR gets merged?

To do that, you just need to add the following like to the end of the PR description:

Closes https://github.com/keycloak/keycloak/issues/30211

**Comment by MaciejMierzwa** [comment]
HI, I have modified the tests. Hopefully, the tests will be stable now.

**Comment by MaciejMierzwa** [comment]
Hi @pedroigor, could you trigger CI?

**Comment by MaciejMierzwa** [comment]
Hi everybody, imo the feature is ready to go. It looks like all tests are green. One issue I have is that I changed documentation which triggers the documentation validation job. There are some expired links, but it's not in caused by my changes. Looks like it's the scope of a different task.
Could you please advise on that?
@sguilhen 
@pedroigor 
@thomasdarimont 
@mabartos

**Comment by MaciejMierzwa** [comment]
Hi, looks like the pipeline is green. My tests are run in job: Keycloak CI / Base IT (6)

**Comment by MaciejMierzwa** [comment]
Hi, are there any additional steps I have to take to get this merged/released? No additional activity from me, except rebasing.
@sguilhen
@pedroigor
@thomasdarimont
@mabartos

File: server-spi-private/src/main/java/org/keycloak/policy/AgePasswordPolicyProvider.java
Comment (sguilhen) [inline server-spi-private/src/main/java/org/keycloak/policy/AgePasswordPolicyProvider.java:45]: It looks like this method is ultimately called only in tests, but wouldn't it be better to invoke the other variant from here instead of returning `null`? Something like

```
RealmModel realm = session.getContext().getRealm();
return validate(realm, session.users().getUserByUsername(realm, user), password);
```
Just a suggestion, I wouldn't reject the PR because of this.

Comment (MaciejMierzwa) [reply server-spi-private/src/main/java/org/keycloak/policy/AgePasswordPolicyProvider.java]: Def agreed, pushed changes with your suggestion

**Comment by MaciejMierzwa** [comment]
Hi, looks like the feature works, all tests are passing, there is documentation ready in PR, and the only issue is the amount of changes coming to the main branch and rebasing it. 
What should I do to get this merged/released?
@sguilhen
@pedroigor
@thomasdarimont
@mabartos

**Comment by MaciejMierzwa** [comment]
Hi @sguilhen pinging for visibility, I saw you assigned this task to yourself. I implemented the changes you suggested: https://github.com/keycloak/keycloak/pull/30215#discussion_r1669126023

**Comment by sguilhen** [comment]
Hi @MaciejMierzwa - yeah, I assigned it to myself to ensure I can help drive this to completion. I've already asked for maintainer approval for the PR, I'm hoping to get this in shortly

**Review by pedroigor — APPROVED** [review]
@MaciejMierzwa Thanks for your time. I have just updated the docs to make it a bit more clear about the intention of the new policy.

Please, let me know what you think. LGTM.

**Pull request merged**
